state-commitments/optimized/history_commitment.go

```diff
@@ -24,17 +24,46 @@ func NewCommitment(leaves []common.Hash, virtual uint64) (*Commitment, error) {
 	if len(leaves) == 0 {
 		return nil, errors.New("must commit to at least one leaf")
 	}
+	if virtual == 0 {
+		return nil, errors.New("virtual size cannot be zero")
+	}
 	comm := NewCommitter()
-	root, err := comm.ComputeRoot(leaves, virtual)
+	leavesForProofs := make([]common.Hash, len(leaves))
+	leavesForCommit := make([]common.Hash, len(leaves))
+	copy(leavesForProofs, leaves)
+	firstLeaf := leavesForProofs[0]
+	lastLeaf := leavesForProofs[len(leavesForProofs)-1]
+	for i := 0; i < len(leavesForProofs); i++ {
+		result, err := comm.hash(leavesForProofs[i][:])
+		if err != nil {
+			return nil, err
+		}
+		leavesForCommit[i] = result
+		leavesForProofs[i] = result
```

Comment:
We shouldn't copy  `leaves` if we're overwriting the copy right here unconditionally